### ISSUE TITLE
ClassPageRenderer takes AGES because it double-calls an expensive method

### ISSUE THREAD

**Issue by shevek** [gradle/gradle#10045]
* We have a parameterized test with >50,000 cases.
* The test has a Parameterized.Parameters(name= annotation) which I think causes TestResult.name == TestResult.displayName
* Generating the HTML report for a test result takes AGES.
* The time is spent in ClassPageRenderer.methodNameColumnExists
* This is because the method is called TWICE:
** Once in methodNameColumnExists -> determineTableHeaders()
** Once in methodNameColumnExists() directly.

### Expected Behavior
Call the method only once and pass the boolean to determineTableHeaders, that would at least double the performance.

### Current Behavior
Our test engineers get bored waiting for the pages to render, and go to the pub.

### Context
All our test engineers are in the pub because generating HTML reports takes so damn long.

### Steps to Reproduce 
A junit test with 50,000 cases, for each of which TestResult.getName() equals TestResult.getDisplayName() - 

### Your Environment
Nearest pub: 2 minutes away.
Time taken to run 50,000 test: 3 minutes.
Time taken to render page results: 4 minutes.
Time spent drinking = (3+4 [test+result]) - (2+2 [walk there and back]) = 3 minutes.

If we can get the drinking time down to 1 minute per test run, we can probably make it to 5pm without actually getting blind drunk.

I mean, it really ought to be faster to render a table result row faster than it's possible to run an entire unit test which, in many cases, invokes a compiler!

**Comment by mattjquinn** [comment]
Similar to Shevek, I also experience this issue myself, in a situation very much like the one he himself describes. Gradle is quite a wonderful build system overall, but after all the trips to the pub I have experienced while waiting for my HTML test report to generate, I am becoming an alcoholic.

**Comment by vlsi** [comment]
@shevek , @mattjquinn , have you tried Gradle 5.5.1?
The issue seems to be fixed already: https://github.com/gradle/gradle/commit/a5436a4681109a2dc91892a2a3ee4223d6a3a70a

**Comment by shevek** [comment]
@vlsi  Nope, I just looked at the hashref you posted, and it contains the EXACT buggy callchain described in this issue: Two calls to methodNameColumnExists().

* https://github.com/gradle/gradle/blob/a5436a4681109a2dc91892a2a3ee4223d6a3a70a/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/ClassPageRenderer.java#L54
* https://github.com/gradle/gradle/blob/a5436a4681109a2dc91892a2a3ee4223d6a3a70a/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/ClassPageRenderer.java#L69

So this issue is NOT fixed.

**Comment by vlsi** [comment]
@shevek ,  Even though the method is executed twice, it would not make much difference.

Can you provide a reproducer?

I'm quite sure current Gradle code would not spend 4 minutes in the methods in `methodNameColumnExists`

**Comment by shevek** [comment]
I can provide you a perf trace showing that it does.

**Comment by vlsi** [comment]
Please do so. Even better would be if you could provide a project or unit-test to reproduce the behaviour.

**Comment by shevek** [comment]
Confirmed, jstack says it spent 2m30s in:
```
   java.lang.Thread.State: RUNNABLE
        at org.gradle.api.internal.tasks.testing.report.ClassPageRenderer.methodNameColumnExists(ClassPageRenderer.java:91)
        at org.gradle.api.internal.tasks.testing.report.ClassPageRenderer.determineTableRow(ClassPageRenderer.java:61)
        at org.gradle.api.internal.tasks.testing.report.ClassPageRenderer.renderTableRow(ClassPageRenderer.java:80)
        at org.gradle.api.internal.tasks.testing.report.ClassPageRenderer.renderTests(ClassPageRenderer.java:55)
        at org.gradle.api.internal.tasks.testing.report.ClassPageRenderer.access$000(ClassPageRenderer.java:32)
        at org.gradle.api.internal.tasks.testing.report.ClassPageRenderer$1.doExecute(ClassPageRenderer.java:119)
```

Here is 60 seconds of perf trace, which shows a slightly different answer; bear in mind that the horizontal width here represents 60 seconds out of the total 140-ish seconds of this call.
![gradle-render-table-row](https://user-images.githubusercontent.com/391350/62066172-6f090500-b1e5-11e9-97b1-d0d61c01bff6.png) The double green box top right is methodNameColumnExists. I would upload the interactive SVG if github would let me.

methodNameColumnExists MIGHT be slow because the underlying Map is a TreeMap, which is slow as a dog. I have no hypothesis about why renderTableRow is so slow.

Please fix this, at least part of the fix is trivial, and it takes FAR too much of our developers' time waiting for this.

**Comment by shevek** [comment]
I suspect if I grab a full 3 minute trace, the situation will appear worse, which it is.

**Comment by vlsi** [comment]
@shevek , would you please capture results via https://github.com/jvm-profiling-tools/async-profiler ?

**Comment by vlsi** [comment]
@shevek , would you please clarify which Gradle version do you use?
Please upgrade to Gradle 5.2.0 or later and close the issue.

There's nothing to do for Gradle team here.


>shevek: Confirmed, jstack says it spent 2m30s in:

```
   java.lang.Thread.State: RUNNABLE
        at org.gradle.api.internal.tasks.testing.report.ClassPageRenderer.
methodNameColumnExists(ClassPageRenderer.java:91)
```

It is obvious that you are using **OLD** Gradle version, because line 91 is just EMPTY now:

https://github.com/gradle/gradle/blob/a5436a4681109a2dc91892a2a3ee4223d6a3a70a/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/ClassPageRenderer.java#L61

**Comment by shevek** [comment]
The bug is STILL PRESENT in 5.2.0. Upgrading will do nothing for me until you fix the bug. The post I made was already done using perf, which is already async, and not safepoint-sensitive.

**Comment by vlsi** [comment]
> The bug is STILL PRESENT in 5.2.0

I'm afraid you are wrong.

Here's a direct URL for version 5.2.0: https://github.com/gradle/gradle/blob/v5.2.0/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/ClassPageRenderer.java#L61

The line 61 is empty.

The issue has already been fixed, so you just need to upgrade and that is it.

**Comment by shevek** [comment]
@vlsi Nope:

Call 1: https://github.com/gradle/gradle/blob/v5.2.0/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/ClassPageRenderer.java#L69
Call 2: https://github.com/gradle/gradle/blob/v5.2.0/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/ClassPageRenderer.java#L54

Both called from https://github.com/gradle/gradle/blob/v5.2.0/subprojects/testing-base/src/main/java/org/gradle/api/internal/tasks/testing/report/ClassPageRenderer.java#L50

Still present in 5.2.0, exactly as described above.

**Comment by vlsi** [comment]
> Still present in 5.2.0, exactly as described above.

Please upgrade to Gradle 5.2.0 (or later) and provide performace traces (e.g. async-profiler or jstack).

Your current analysis shows that you are using an outdated version, and you somehow fail to realize that what you need is https://github.com/gradle/gradle/commit/a5436a4681109a2dc91892a2a3ee4223d6a3a70a which comes in Gradle 5.2.0+
